File: packages/nativescript/platforms/ios/core/NPFileDelta.swift
import Foundation

/// Bounded NPDIFF01 decoder. File hashes and signatures are verified by the caller.
enum NPFileDelta {
    static func apply(base: Data, patch: Data, maximumBytes: Int) throws -> Data {
        func invalid() -> NSError { NSError(domain: "NitroPushFileDelta", code: 1) }
        guard !base.isEmpty, base.count <= maximumBytes, patch.count >= 12,
              patch.count <= maximumBytes, patch.prefix(8) == Data("NPDIFF01".utf8) else { throw invalid() }
        var cursor = 8
        func u32() throws -> Int {
            guard cursor + 4 <= patch.count else { throw invalid() }
            var result = 0
            for _ in 0..<4 { result = (result << 8) | Int(patch[cursor]); cursor += 1 }
            return result
        }
        let size = try u32()
        guard size > 0, size <= maximumBytes else { throw invalid() }
        var output = Data(); output.reserveCapacity(size)
        while cursor < patch.count {
            let op = patch[cursor]; cursor += 1
            guard op == 0 || op == 1 else { throw invalid() }
            let offset = op == 0 ? try u32() : 0
            let length = try u32()
            guard length > 0, length <= size - output.count else { throw invalid() }
            if op == 0 {
                guard offset <= base.count, length <= base.count - offset else { throw invalid() }
                output.append(base.subdata(in: offset..<(offset + length)))
            } else {
                guard length <= patch.count - cursor else { throw invalid() }
                output.append(patch.subdata(in: cursor..<(cursor + length))); cursor += length
            }
        }
        guard output.count == size else { throw invalid() }
        return output
    }
}

File: packages/nativescript/platforms/android/java/com/nitropush/sdk/NPFileDelta.kt
package com.nitropush.sdk

import java.io.ByteArrayOutputStream

/** Bounded NPDIFF01 decoder. The caller authenticates base and output bytes. */
internal object NPFileDelta {
    fun apply(base: ByteArray, patch: ByteArray, maximumBytes: Int): ByteArray {
        check(base.isNotEmpty() && base.size <= maximumBytes && patch.size in 12..maximumBytes &&
            patch.copyOfRange(0, 8).contentEquals("NPDIFF01".toByteArray(Charsets.US_ASCII))) { "Invalid file delta header" }
        var cursor = 8
        fun u32(): Int {
            check(cursor + 4 <= patch.size) { "Truncated file delta" }
            var value = 0L
            repeat(4) { value = (value shl 8) or (patch[cursor++].toLong() and 255) }
            check(value <= Int.MAX_VALUE) { "File delta integer exceeds limit" }
            return value.toInt()
        }
        val size = u32()
        check(size in 1..maximumBytes) { "Invalid file delta output size" }
        val output = ByteArrayOutputStream(size)
        while (cursor < patch.size) {
            val op = patch[cursor++].toInt()
            check(op == 0 || op == 1) { "Invalid file delta operation" }
            val offset = if (op == 0) u32() else 0
            val length = u32()
            check(length > 0 && length <= size - output.size()) { "File delta exceeds output size" }
            if (op == 0) {
                check(offset <= base.size && length <= base.size - offset) { "File delta exceeds base size" }
                output.write(base, offset, length)
            } else {
                check(length <= patch.size - cursor) { "Truncated file delta literal" }
                output.write(patch, cursor, length); cursor += length
            }
        }
        check(output.size() == size) { "Incomplete file delta" }
        return output.toByteArray()
    }
}
